satellite/metainfo/loop.go

```diff
@@ -0,0 +1,252 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package metainfo
+
+import (
+	"context"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storage"
+)
+
+var (
+	// LoopError is a standard error class for this component.
+	LoopError = errs.Class("metainfo loop error")
+	// LoopClosedError is a loop closed error
+	LoopClosedError = LoopError.New("loop closed")
+)
+
+// Observer is an interface defining an observer that can subscribe to the metainfo loop.
+type Observer interface {
+	RemoteSegment(context.Context, storj.Path, *pb.Pointer) error
+	RemoteObject(context.Context, storj.Path, *pb.Pointer) error
+	InlineSegment(context.Context, storj.Path, *pb.Pointer) error
+}
+
+type observerContext struct {
+	Observer
+	ctx  context.Context
+	done chan error
+}
+
+func (observer *observerContext) HandleError(err error) bool {
+	if err != nil {
+		observer.done <- err
+		observer.Finish()
+		return true
+	}
+	return false
+}
+
+func (observer *observerContext) Finish() {
+	close(observer.done)
+}
+
+func (observer *observerContext) Wait() error {
+	return <-observer.done
+}
+
+// LoopConfig contains configurable values for the metainfo loop.
+type LoopConfig struct {
+	CoalesceDuration time.Duration `help:"how long to wait for new observers before starting iteration" releaseDefault:"5s" devDefault:"5s"`
+}
+
+// Loop is a metainfo loop service.
+type Loop struct {
+	config   LoopConfig
+	metainfo *Service
+	join     chan *observerContext
+	// "done" is closed when loop.Run completely exits
+	done chan struct{}
+	// "close" is closed when loop.Close() is called
+	close chan struct{}
+}
+
+// NewLoop creates a new metainfo loop service.
+func NewLoop(config LoopConfig, metainfo *Service) *Loop {
+	return &Loop{
+		metainfo: metainfo,
+		config:   config,
+		join:     make(chan *observerContext),
+		done:     make(chan struct{}),
+		close:    make(chan struct{}),
```

Comment:
this extra close channel shouldn't be necessary

Reply:
sorry didn't have my latest changes pushed